components/Banner/examples/supplemental/MDCBannerView.m

```diff
@@ -54,124 +70,330 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
-  return self.buttons.count;
-}
+// TODO: KVO trailingButton.hidden and call `setNeedsUpdateConstraints:`.
 
 #pragma mark - UIView overrides
 
 - (CGSize)sizeThatFits:(CGSize)size {
-  self.layout = [[MDCBannerViewLayout alloc] initWithSizeToFit:size
-                                                     textLabel:self.textLabel
-                                                imageContainer:self.iconImageViewContainer
-                                                       buttons:self.buttons];
-  return [self.layout frameSize];
+  MDCBannerViewLayoutStyle style = [self styleForSizeToFit:size];
+  CGFloat frameHeight = 0.0f;
+  switch (style) {
+    case MDCBannerViewLayoutSingleLineStyle: {
+      frameHeight += kTopPaddingSmall + kBottomPadding;
+      CGSize textLabelSize = [self.textLabel sizeThatFits:CGSizeZero];
+      CGSize leadingButtonSize = [self.leadingButton sizeThatFits:CGSizeZero];
+      CGFloat maximumHeight = MAX(textLabelSize.height, leadingButtonSize.height);
+      if (self.hasImage) {
+        maximumHeight = MAX(kImageViewSideLength, maximumHeight);
+      }
+      frameHeight += maximumHeight;
+      break;
+    }
+    case MDCBannerViewLayoutMultiLineAlignedButtonStyle: {
+      frameHeight += kTopPaddingLarge + kBottomPadding;
+      frameHeight += [self getFrameHeightOfImageViewAndTextLabelWithSizeToFit:size];
+      CGSize leadingButtonSize = [self.leadingButton sizeThatFits:CGSizeZero];
+      CGSize trailingButtonSize = [self.trailingButton sizeThatFits:CGSizeZero];
+      frameHeight += MAX(leadingButtonSize.height, trailingButtonSize.height);
+      break;
+    }
+    case MDCBannerViewLayoutMultiLineStackedButtonStyle: {
+      frameHeight += kTopPaddingLarge + kBottomPadding;
+      frameHeight += [self getFrameHeightOfImageViewAndTextLabelWithSizeToFit:size];
+      CGSize leadingButtonSize = [self.leadingButton sizeThatFits:CGSizeZero];
+      CGSize trailingButtonSize = [self.trailingButton sizeThatFits:CGSizeZero];
+      frameHeight +=
+          leadingButtonSize.height + trailingButtonSize.height + kButtonVerticalIntervalSpace;
+      break;
+    }
+  }
+  return CGSizeMake(size.width, frameHeight);
 }
 
-- (void)layoutSubviews {
-  [super layoutSubviews];
+- (void)updateConstraints {
+  MDCBannerViewLayoutStyle style = [self styleForSizeToFit:self.bounds.size];
+  [self updateConstraintsWithLayoutStyle:style];
 
-  [self applyMDCBannerViewLayout:self.layout];
+  [super updateConstraints];
 }
 
 #pragma mark - Layout methods
 
-- (void)applyMDCBannerViewLayout:(MDCBannerViewLayout *)layout {
-  if (![self isMDCBannerViewLayoutApplicable:layout]) {
-    return;
+- (void)updateConstraintsWithLayoutStyle:(MDCBannerViewLayoutStyle)layoutStyle {
+  // Set Elements
+  if (@available(iOS 9.0, *)) {
+    if (layoutStyle == MDCBannerViewLayoutSingleLineStyle) {
+      if (self.hasImage) {
+        [self.imageView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor
```

Comment:
Is it possible/reasonable to initialize these constraints elsewhere and to then to have this method focus on enabling/disabling the necessary groups of constraints? I think that may simplify the logic here.

Reply:
Good idea. I will refactor this part and try to see if I can make it clean.

Reply:
Addressed in  63130a8